src/System.Net.Security/src/System/Net/_SslSessionsCache.cs

```diff
@@ -0,0 +1,201 @@
+                    if (1 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[1] << 8);
+                    if (2 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[2] << 16);
+                    if (3 < _CertThumbPrint.Length)
+                        _HashCode ^= (_CertThumbPrint[3] << 24);
+                }
+                _HashCode ^= allowedProtocols;
+                _HashCode ^= (int)encryptionPolicy;
+                _AllowedProtocols = allowedProtocols;
+                _EncryptionPolicy = encryptionPolicy;
+            }
+
+            public override int GetHashCode()
+            {
+                return _HashCode;
+            }
+
+            public static bool operator ==(SslCredKey sslCredKey1,
+                                            SslCredKey sslCredKey2)
+            {
+                return sslCredKey1.Equals(sslCredKey2);
+            }
+
+            public static bool operator !=(SslCredKey sslCredKey1,
+                                            SslCredKey sslCredKey2)
+            {
+                return !sslCredKey1.Equals(sslCredKey2);
+            }
+
+            public override bool Equals(Object y)
+            {
+                SslCredKey other = (SslCredKey)y;
+
+                if (_CertThumbPrint.Length != other._CertThumbPrint.Length)
+                    return false;
+
+                if (_HashCode != other._HashCode)
+                    return false;
+
+                if (_EncryptionPolicy != other._EncryptionPolicy)
+                    return false;
+
+                if (_AllowedProtocols != other._AllowedProtocols)
+                    return false;
+
+                for (int i = 0; i < _CertThumbPrint.Length; ++i)
+                    if (_CertThumbPrint[i] != other._CertThumbPrint[i])
+                        return false;
+
+                return true;
+            }
+        }
+
+        //
+        // Returns null or previously cached cred handle.
+        //
+        // ATTN: The returned handle can be invalid, the callers of InitializeSecurityContext and AcceptSecurityContext
+        // must be prepared to execute a back-out code if the call fails.
+        //
+        internal static SafeFreeCredentials TryCachedCredential(byte[] thumbPrint, int allowedProtocols, EncryptionPolicy encryptionPolicy)
+        {
+            if (s_CachedCreds.Count == 0)
+            {
+                GlobalLog.Print("TryCachedCredential() Not Found, Current Cache Count = " + s_CachedCreds.Count);
+                return null;
+            }
+
+            object key = new SslCredKey(thumbPrint, allowedProtocols, encryptionPolicy);
+
+            SafeCredentialReference cached = s_CachedCreds[key] as SafeCredentialReference;
+
+            if (cached == null || cached.IsClosed || cached._Target.IsInvalid)
+            {
+                GlobalLog.Print("TryCachedCredential() Not Found, Current Cache Count = " + s_CachedCreds.Count);
```

Comment:
The logging message here doesn't really match the condition... if it was closed out it was "found", just not "usable"...